### PR TITLE
fix: circular deps finalized head

### PR DESCRIPTION
**Pull request by parketh**
## Summary

This PR addresses various circular dependency issues with the current FG implementation.

### Problems

- FG currently only processes blocks [above the finalized head](https://github.com/babylonlabs-io/finality-gadget/blob/main/finalitygadget/finalitygadget.go#L432). However, the finalized head must be both btc+eth finalized. This means the FG never advances past the eth finalized block (it never achieves fast finality). 
- Basing the block processing logic on the RPC also introduces potential circular dependencies
- block processing is currently very slow as it is sequential. if the FG falls behind for any reason (e.g. crash in another system service), it may never catch up (we observe this in our devnet)

### Proposed fixes

1. Decouple the startup loop from the block processing loop. Only use the eth finalized block on startup, not anywhere else in the code.
2. Implement batch insert by parallelising calls
3. Update `QueryBlockRangeBabylonFinalized` to query the internal DB rather than call the Babylon node

## Test plan

```
make lint
make test
```

### CODE REVIEW

File: config.toml.example [hunk 1]
```diff
@@ -9,4 +9,5 @@ BBNChainID = "euphrates-0.4.0"
 BBNRPCAddress = "https://rpc-euphrates.devnet.babylonchain.io"
 GRPCListener = "0.0.0.0:50051"
 HTTPListener = "0.0.0.0:8080"
-PollInterval = "10s"
\ No newline at end of file
+PollInterval = "10s"
+BatchSize = 10
```
Comment (bap2pecs) [inline config.toml.example:13]: is this only used for fast sync? if that's the case, calling it FastSyncBatchSize would be better

File: finalitygadget/finalitygadget.go [hunk 2]
```diff
@@ -384,8 +385,14 @@ func (fg *FinalityGadget) QueryLatestFinalizedBlock() (*types.Block, error) {
 	return fg.db.QueryLatestFinalizedBlock()
 }
 
-// This function process blocks indefinitely, starting from the last finalized block.
-func (fg *FinalityGadget) ProcessBlocks(ctx context.Context) error {
+// This function is run once at startup and starts the FG from the last finalized block.
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:388]: if this is run once, do we still need the logic "Start polling for new blocks at set interval" below?

File: finalitygadget/interface.go [hunk 3]
```diff
@@ -3,7 +3,7 @@ package finalitygadget
 import "github.com/babylonlabs-io/finality-gadget/types"
 
 type IFinalityGadget interface {
-	/* QueryIsBlockBabylonFinalized checks if the given L2 block is finalized by the Babylon finality gadget
+	/* QueryIsBlockBabylonFinalizedFromBabylon checks if the given L2 block is finalized by the Babylon finality gadget
```
Comment (bap2pecs) [inline finalitygadget/interface.go:6]: what's the difference w the old name?

File: config.toml.example (hunk 1 above)
Comment (bap2pecs) [reply config.toml.example:13]: discussed offline. this is not really fast sync

File: finalitygadget/interface.go (hunk 3 above)
Comment (bap2pecs) [reply finalitygadget/interface.go:6]: ok it seem `QueryIsBlockBabylonFinalizedFromBabylon` calculates quorum on the fly while `QueryBlockRangeBabylonFinalized` read from db

so I think for all external RPCs calls, we should always read from DB to avoid DoS attack.

in that case, should we remove the RPC?

File: finalitygadget/finalitygadget.go (hunk 2 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go:388]: yes both loops are necessary:
- startup loop runs until btc staking is activated
- block processing loop runs to fetch new blocks

File: finalitygadget/interface.go (hunk 3 above)
Comment (parketh) [reply finalitygadget/interface.go:6]: we can make `queryIsBlockBabylonFinalizedFromBabylon` internal to prevent unwanted calls. i left it external for now as we may need it for testing (e.g. if FG returns block not finalized, we would want to query Babylon node directly to debug)

File: config.toml.example (hunk 1 above)
Comment (parketh) [reply config.toml.example:13]: yep, to disable batch polling / inserts, consumer can set `BatchSize` to 1

File: finalitygadget/finalitygadget.go (hunk 2 above)
Comment (bap2pecs) [reply finalitygadget/finalitygadget.go:388]: got it. so the `ProcessBlocks` go routine won't start until btc staking is activated?

File: finalitygadget/interface.go (hunk 3 above)
Comment (bap2pecs) [reply finalitygadget/interface.go:6]: sure we can make it external for now for development. maybe add a TODO there so remove it in the future once everything is stable?

**Comment by lesterli** [comment]
for this change, we need to update the FP e2e test. 
I just ran `make test-e2e-op` with this commit and new rpc `QueryIsBlockBabylonFinalizedFromBabylon`, got this error:
```
2024-11-02T12:58:34.164+0800    ERROR   finalitygadget/finalitygadget.go:275    Failed to get activated timestamp from database {"error": "database not open"}
github.com/babylonlabs-io/finality-gadget/finalitygadget.(*FinalityGadget).QueryBtcStakingActivatedTimestamp
        /Users/lyy/go/pkg/mod/github.com/babylonlabs-io/finality-gadget@v0.1.2-0.20241101114209-e60a22e09820/finalitygadget/finalitygadget.go:275
github.com/babylonlabs-io/finality-gadget/server.(*rpcServer).QueryBtcStakingActivatedTimestamp
        /Users/lyy/go/pkg/mod/github.com/babylonlabs-io/finality-gadget@v0.1.2-0.20241101114209-e60a22e09820/server/rpcserver.go:80
github.com/babylonlabs-io/finality-gadget/proto._FinalityGadget_QueryBtcStakingActivatedTimestamp_Handler
        /Users/lyy/go/pkg/mod/github.com/babylonlabs-io/finality-gadget@v0.1.2-0.20241101114209-e60a22e09820/proto/finalitygadget_grpc.pb.go:218
google.golang.org/grpc.(*Server).processUnaryRPC
        /Users/lyy/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1379
google.golang.org/grpc.(*Server).handleStream
        /Users/lyy/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1790
google.golang.org/grpc.(*Server).serveStreams.func2.1
        /Users/lyy/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1029
```

**Comment by bap2pecs** [comment]
> I just ran `make test-e2e-op` with this commit and new rpc `QueryIsBlockBabylonFinalizedFromBabylon`, got this error:

I also met with this error testing my PR. i believe it's b/c `TestFinalityProviderLifeCycle` is broken already

**Comment by parketh** [comment]
> > I just ran `make test-e2e-op` with this commit and new rpc `QueryIsBlockBabylonFinalizedFromBabylon`, got this error:
> 
> I also met with this error testing my PR. the problem happens without this PR. i think it's likely caused by one of the commit here:
> 
> image
> I guess this also means whenever we have a FG PR, we should test it with FP or it will be hard to bisect :)

i narrowed it down to [this](https://github.com/babylonlabs-io/finality-gadget/pull/24/commits/121402c6573f0829fe4759c6ae53696f4aaf755d#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6) commit. It seems adding the http server breaks something, but cant figure out after debugging for hours..

weirdly this error doesnt come up in prod, only e2e tests

**Comment by parketh** [comment]
another observation - running the tests individually e.g. `make test-e2e-op-filter FILTER=TestFinalityProviderLifeCycle` causes the error to disappear. this seems to suggest some kind of race condition, but its not clear how this is related to the http server

File: server/rpcserver.go [hunk 4]
```diff
@@ -35,9 +35,9 @@ func (r *rpcServer) RegisterWithGrpcServer(grpcServer *grpc.Server) error {
 	return nil
 }
 
-// QueryIsBlockBabylonFinalized is an RPC method that returns the finality status of a block by querying Babylon chain.
-func (r *rpcServer) QueryIsBlockBabylonFinalized(ctx context.Context, req *proto.QueryIsBlockBabylonFinalizedRequest) (*proto.QueryIsBlockFinalizedResponse, error) {
```
Comment (bap2pecs) [inline server/rpcserver.go:39]: do we have any external dependencies (e.g. op-node, finality-explorer, FP) that depends on this function?

File: go.mod [hunk 5]
```diff
@@ -15,7 +15,7 @@ require (
 	github.com/ethereum/go-ethereum v1.13.15
 	github.com/jsternberg/zap-logfmt v1.3.0
 	github.com/lightningnetwork/lnd v0.16.4-beta.rc1
-	github.com/rs/cors v1.11.0
+	github.com/rs/cors v1.8.3
```
Comment (bap2pecs) [inline go.mod:18]: why do we need a downgrade?

File: finalitygadget/finalitygadget.go [hunk 6]
```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
 	}, nil
 }
 
-func (fg *FinalityGadget) handleBlock(ctx context.Context, latestFinalizedHeight uint64) error {
-	for height := fg.lastProcessedHeight + 1; height <= latestFinalizedHeight; height++ {
+type blockResult struct {
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:577]: i had a hard time following this struct and related logic. my understanding:

- `blockResult` is mostly only used as the return type for `processHeight`
- `processHeight` is only called by `processBlocksTillHeight`
- the `block` field is never used if it's not finalized

I think we can completely get rid of this struct to make the logic cleaner:
- refactor `processHeight` to `processBlock(block *types.Block) (error, bool)`
- if we don't want to add another go routine for batch calling `queryBlockByHeight()`, we can refactor the function to `processHeightAndReturnIfFinalized(height uint64) (error, *types.Block)`
- if it's not finalized, it will return `nil, nil`

File: server/rpcserver.go (hunk 4 above)
Comment (bap2pecs) [reply server/rpcserver.go:39]: why are we not adding a new API but renaming the old one?

i see in `finalitygadget/interface.go` we added a new API. shouldn't we match it here?

File: finalitygadget/finalitygadget.go [hunk 7]
```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
...
+					processingError = result.err
+					break
+				}
+				statusMap[result.height] = result
 			}
-			// If not finalized, throw error
-			if !isFinal {
-				return fmt.Errorf("block %d should be finalized according to client but is not", block.BlockHeight)
+			if processingError != nil {
+				return processingError
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:630]: in the goroutine processing heights above, we could add a context with cancel to cancel all goroutines as long as there is one goroutine finished with error

but it's probably a low-pri. we can just add a TODO comment

File: finalitygadget/finalitygadget.go [hunk 8]
```diff
@@ -207,8 +210,38 @@ func (fg *FinalityGadget) QueryIsBlockBabylonFinalized(block *types.Block) (bool
...
+	}
+
+	// check whether the btc staking is activated
+	btcStakingActivatedTimestamp, err := fg.QueryBtcStakingActivatedTimestamp()
+	if err != nil {
+		return false, err
+	}
+	if btcblockHeight < btcStakingActivatedTimestamp {
+		return false, types.ErrBtcStakingNotActivated
+	}
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:238]: i think there is no need to check this. we only insert blocks to the db after BTC finalization.

this function should simply read from DB

it's also cleaner so we don't need to copy the above code

File: server/rpcserver.go (hunk 4 above)
Comment (parketh) [reply server/rpcserver.go:39]: yes op e2e tests in fp repo call this function. the changes are:

- update `QueryIsBlockBabylonFinalized` to now call internal db
- move old logic to a new `QueryIsBlockBabylonFinalizedFromBabylon` function, which we can keep using for testing purposes, but wont be called by external services

File: finalitygadget/finalitygadget.go [hunk 9]
```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
...
-	for height := fg.lastProcessedHeight + 1; height <= latestFinalizedHeight; height++ {
+type blockResult struct {
+	height      uint64
+	block       *types.Block
+	isFinalized bool
+	err         error
+}
+
+func (fg *FinalityGadget) processBlocksTillHeight(ctx context.Context, latestHeight uint64) error {
+	for startHeight := fg.lastProcessedHeight + 1; startHeight <= latestHeight; {
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:585]: `startHeight` sounds like at the level of the whole function. i was gonna suggest to rename to `currentHeight` but then I realize it's for a batch.

so `batchStartHeight` and `batchEndHeight` seems to be better names

File: finalitygadget/finalitygadget.go [hunk 10]
```diff
@@ -384,8 +424,14 @@ func (fg *FinalityGadget) QueryLatestFinalizedBlock() (*types.Block, error) {
 	return fg.db.QueryLatestFinalizedBlock()
 }
 
-// This function process blocks indefinitely, starting from the last finalized block.
-func (fg *FinalityGadget) ProcessBlocks(ctx context.Context) error {
+// This function is run once at startup and starts the FG from the last finalized block.
+// Note that this logic will fail if the FG is started before the chain has ETH finalized its first block.
+// The intended startup order for new chains is:
+//  1. Integrate FG with it disabled on CW contract
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:430]: after my new PR in the OP repo, the OP chain will always be up first. so the order should be:

1. Start the OP chain with `babylonFinalityGadgetRpc` rollup.config set to empty string. wait for it to have at least 1 eth-finalized block
2. ...
3. restart `op-node` after setting `babylonFinalityGadgetRpc` with the FG gRPC url
4. enable finality gadget in the cw contract

File: finalitygadget/finalitygadget.go [hunk 11]
```diff
@@ -413,47 +460,75 @@ func (fg *FinalityGadget) ProcessBlocks(ctx context.Context) error {
 				return fmt.Errorf("error querying BTC staking activation timestamp: %w", err)
 			}
 
-			// only process blocks after the btc staking is activated
+			// throw error if btc staking activated before the first block was finalized (see startup order above)
+			if latestFinalizedHeight == 0 && latestFinalizedBlockTime < btcStakingActivatedTimestamp {
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:464]: shouldn't the logic be 

btcStakingActivatedTimestamp < latestFinalizedBlockTime

File: finalitygadget/finalitygadget.go [hunk 12]
```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
...
+			fg.logger.Info("Processing batch of blocks", zap.Uint64("start_height", startHeight), zap.Uint64("end_height", endHeight))
+
+			// Create batch of blocks to check in parallel
+			results := make(chan blockResult, endHeight-startHeight+1)
+			var wg sync.WaitGroup
+
+			// Query batch in parallel
+			for height := startHeight; height <= endHeight; height++ {
+				if height > math.MaxInt64 {
+					return fmt.Errorf("block height %d exceeds maximum int64 value", height)
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:604]: why do we need this check? if there a type conversion below?

but anyway shouldn't we check `endHeight` instead of `startHeight`? if yes, shouldn't we just check the passed in argument `latestHeight` at the first line of the function instead of inside the loop?

File: finalitygadget/finalitygadget.go [hunk 13]
```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
...
+	block, err := fg.queryBlockByHeight(int64(height))
+	if err != nil {
+		result.err = fmt.Errorf("error getting block at height %d: %w", height, err)
+		return result
+	}
+	result.block = block
+
+	// Check finalization
+	isFinalized, err := fg.QueryIsBlockBabylonFinalizedFromBabylon(block)
+	if err != nil && !errors.Is(err, types.ErrBtcStakingNotActivated) {
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget.go:679]: `processBlocksTillHeight` should never be called when btc staking is not activated

so having this check if unnecessary and can cause confusion (i.e. what should the behavior of this function be when it's not btc staking activated)

File: finalitygadget/finalitygadget.go (hunk 6 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go]: updated for this in latest commit

File: server/rpcserver.go (hunk 4 above)
Comment (bap2pecs) [reply server/rpcserver.go:39]: > update QueryIsBlockBabylonFinalized to now call internal db

wait I see `QueryIsBlockBabylonFinalized` is completely removed from `server/rpcserver.go`. Is that right?

shouldn't we have both `QueryIsBlockBabylonFinalizedFromBabylon` and `QueryIsBlockBabylonFinalized` in this file?

File: finalitygadget/finalitygadget.go (hunk 8 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go:238]: we need these checks because the function can be called by op node before btc staking is activated. It should return the correct error `ErrBtcStakingNotActivated` if btc staking is not activated

if we just return the db call, the error type will be incorrect (i think one of the e2e tests broke for this reason)

File: config/config.go [hunk 14]
```diff
@@ -19,6 +20,51 @@ type Config struct {
...
+	}
+	if c.BitcoinRPCHost == "" {
+		return fmt.Errorf("bitcoin-rpc-host is required")
+	}
+	if c.BitcoinRPCUser == "" {
+		return fmt.Errorf("bitcoin-rpc-user is required")
+	}
+	if c.BitcoinRPCPass == "" {
+		return fmt.Errorf("bitcoin-rpc-pass is required")
+	}
```
Comment (bap2pecs) [inline config/config.go:39]: iirc we used the Ankr RPC that doesn't need user and password in our e2e test and it's working

File: config/config.go [hunk 15]
```diff
@@ -19,6 +20,51 @@ type Config struct {
...
+	if c.DBFilePath == "" {
+		return fmt.Errorf("db-file-path is required")
+	}
+	if c.GRPCListener == "" {
+		return fmt.Errorf("grpc-listener is required")
+	}
+	if c.HTTPListener == "" {
+		return fmt.Errorf("http-listener is required")
+	}
```
Comment (bap2pecs) [inline config/config.go:57]: do you think we should assign them some default values if missing? low pri tho

File: finalitygadget/finalitygadget.go (hunk 11 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go]: good spot - updated!

File: config/config.go [hunk 16]
```diff
@@ -19,6 +20,51 @@ type Config struct {
...
+
+	// Numeric validations
+	if c.PollInterval <= 0 {
+		return fmt.Errorf("poll-interval must be positive")
+	}
+	if c.BatchSize == 0 {
```
Comment (bap2pecs) [inline config/config.go:63]: i think we should set a limit. e.g. we don't want to have 1M go routines

File: finalitygadget/finalitygadget.go (hunk 12 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go:600]: this is old code - i think its because eth rpc uses i64 for block heights

File: finalitygadget/finalitygadget.go (hunk 13 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go]: now addressed by latest commit

File: config/config.go (hunk 15 above)
Comment (parketh) [reply config/config.go:52]: i added some todos for now

File: finalitygadget/finalitygadget.go (hunk 8 above)
Comment (bap2pecs) [reply finalitygadget/finalitygadget.go:238]: ah so `QueryIsBlockBabylonFinalized` is not just called by `processHeight`?

File: finalitygadget/finalitygadget.go (hunk 12 above)
Comment (bap2pecs) [reply finalitygadget/finalitygadget.go:600]: yes you can see there is a type conversion later. but my point is this check is incorrect/incomplete.

we can remove it here and add this to the top:
```
func (fg *FinalityGadget) processBlocksTillHeight(ctx context.Context, latestHeight uint64) error {
	if latestHeight > math.MaxInt64 {
		return fmt.Errorf("block height %d exceeds maximum int64 value", height)
	}
```

File: finalitygadget/finalitygadget.go (hunk 8 above)
Comment (parketh) [reply finalitygadget/finalitygadget.go:238]: yep, so:
- `QueryIsBlockBabylonFinalized` is external
- `QueryIsBlockBabylonFinalizedFromBabylon` will eventually be internal (rename `queryIsBlockBabylonFinalizedFromBabylon` once fully tested)

File: db/bbolt.go [hunk 17]
```diff
@@ -64,80 +64,68 @@ func (bb *BBoltHandler) CreateInitialSchema() error {
...
-		return b.Put([]byte(block.BlockHash), bb.itob(block.BlockHeight))
-	})
-	if err != nil {
-		bb.logger.Error("Error inserting block", zap.Error(err))
-		return err
-	}
+	bb.logger.Info("Batch inserting blocks to DB", zap.Int("count", len(blocks)))
+
+	// Single transaction for all operations
+	return bb.db.Update(func(tx *bolt.Tx) error {
```
Comment (bap2pecs) [inline db/bbolt.go:75]: this PR is already too large to review. it's probably better to keep those refactoring stuff in a base PR and have the core logic changes stacked on top of it.

for example, a base PR can merge all the `bb.db.Update` calls inside this function. and other less important stuffs such as adding the config validation logic (that can be its own PR actually)

no need to split anything from this PR tho. just bring up the discussion point

File: db/bbolt.go [hunk 18]
```diff
@@ -64,80 +64,68 @@ func (bb *BBoltHandler) CreateInitialSchema() error {
...
+				minHeight = block.BlockHeight
+			}
+			if block.BlockHeight > maxHeight {
+				maxHeight = block.BlockHeight
+			}
+
+			// Store block data
+			blockBytes, err := json.Marshal(block)
+			if err != nil {
+				return err
```
Comment (bap2pecs) [inline db/bbolt.go:95]: we used to have error logs such as https://github.com/babylonlabs-io/finality-gadget/blob/37a8b1c9b2698c9967bbc060583668d6c7a1e6f9/db/bbolt.go#L81

can we add them back?

File: db/bbolt.go [hunk 19]
```diff
@@ -64,80 +64,68 @@ func (bb *BBoltHandler) CreateInitialSchema() error {
...
-	latestBlock, err := bb.QueryLatestFinalizedBlock()
-	if latestBlock == nil {
-		latestBlock = &types.Block{BlockHeight: 0}
-	}
-	if err != nil {
-		bb.logger.Error("Error getting latest block", zap.Error(err))
-		return err
-	}
+		// Update earliest block if needed
+		earliestBytes := indexBucket.Get([]byte(earliestBlockKey))
```
Comment (bap2pecs) [inline db/bbolt.go:108]: the old code calls `QueryEarliestFinalizedBlock`. is it b/c we don't need to get the actual block so we don't need to call it here?

File: db/bbolt.go [hunk 20]
```diff
@@ -174,6 +162,35 @@ func (bb *BBoltHandler) GetBlockByHash(hash string) (*types.Block, error) {
...
+	len := endHeight - startHeight + 1
+	results := make([]bool, len)
+
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(blocksBucket))
+
+		// Check each height in the range
+		for i := uint64(0); i < len; i++ {
+			height := startHeight + i
+			blockExists := bucket.Get(bb.itob(height)) != nil
```
Comment (bap2pecs) [inline db/bbolt.go:180]: we can probably early return here given the assumption that we only store consecutive finalized blocks. so once we found one slot doesn't exist, we can `return results` directly (b/c all elements in results were initialized to `false` initially)

**Review by bap2pecs — APPROVED** [review]
product management ship it - 9JyQbpKdPa1DeDAFyo

File: db/bbolt.go (hunk 17 above)
Comment (parketh) [reply db/bbolt.go:75]: got it, yeah will try to do this going forward. was prioritising speed but appreciate it prob made code review more difficult!

File: db/bbolt.go (hunk 19 above)
Comment (parketh) [reply db/bbolt.go:108]: yeah its just a more direct way to query. `QueryEarliestFinalizedBlock` is meant to be an external api, so here we can just get the marshalled block directly and check if it exists.

File: .circleci/config.yml
Comment (bap2pecs) [inline .circleci/config.yml:1]: why is this removed? should we add back?

**Comment by bap2pecs** [comment]
lgtm but will let @SebastianElvis take another look

File: finalitygadget/finalitygadget_test.go [hunk 21]
```diff
@@ -17,21 +17,34 @@ import (
...
 )
 
+type TestFinalityGadget struct {
+	*FinalityGadget
+}
+
+// A helper function for testing only
+func (tfg *TestFinalityGadget) InsertBlocks(blocks []*types.Block) error {
+	return tfg.FinalityGadget.insertBlocks(blocks)
+}
```
Comment (bap2pecs) [inline finalitygadget/finalitygadget_test.go:27]: image

we can directly call `insertBlocks` here since it's the same package

Comment (bap2pecs) [reply finalitygadget/finalitygadget_test.go]: fixed in https://github.com/babylonlabs-io/finality-gadget/pull/35/commits/3d85e2f7063fff7f52605e7789027c1b46f0741f

File: db/bbolt.go [hunk 22]
```diff
@@ -174,6 +166,39 @@ func (bb *BBoltHandler) GetBlockByHash(hash string) (*types.Block, error) {
 	return bb.GetBlockByHeight(blockHeight)
 }
 
+func (bb *BBoltHandler) QueryIsBlockRangeFinalizedByHeight(startHeight, endHeight uint64) ([]bool, error) {
```
Comment (SebastianElvis) [inline db/bbolt.go:169]: Would it be possible that block `h` is finalised but block `h+1` is not? I think it's not possiible as BTC staking enforces consecutive finalisation. So wdyt of maintaining the last finalised height in DB and accessing that value, instead of iterating over all these blocks?

**Review by SebastianElvis — APPROVED** [review]
Great work! Some non-blocking comments

File: db/bbolt.go (hunk 22 above)
Comment (bap2pecs) [reply db/bbolt.go:169]: > Would it be possible that block h is finalised but block h+1 is not?

no it's not possible

> So wdyt of maintaining the last finalised height in DB and accessing that value, instead of iterating over all these blocks?

good point. we already maintain fist and last finalized heights. we can use them

```
const (
	blocksBucket          = "blocks"
	blockHeightsBucket    = "block_heights"
	indexerBucket         = "indexer"
	earliestBlockKey      = "earliest"
	latestBlockKey        = "latest"
	activatedTimestampKey = "activated_timestamp"
)
```

Comment (SebastianElvis) [reply db/bbolt.go:169]: Yes, if using last finalised height then the logic can be greatly simplified. Can be done in a separate PR though

**Comment by parketh** [comment]
we tested it with a new devnet and it works! will merge to main
